src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php

```diff
@@ -538,14 +539,26 @@ private function parseDefinition(string $id, $service, string $file, array $defa
             $definition->addTag($name, $tag);
         }
 
-        if (isset($service['decorates'])) {
-            if ('' !== $service['decorates'] && '@' === $service['decorates'][0]) {
-                throw new InvalidArgumentException(sprintf('The value of the "decorates" option for the "%s" service must be the id of the service without the "@" prefix (replace "%s" with "%s").', $id, $service['decorates'], substr($service['decorates'], 1)));
+        if ($decorates = $service['decorates'] ?? null) {
+            if ('' !== $decorates && '@' === $decorates[0]) {
+                throw new InvalidArgumentException(sprintf('The value of the "decorates" option for the "%s" service must be the id of the service without the "@" prefix (replace "%s" with "%s").', $id, $service['decorates'], substr($decorates, 1)));
+            }
+
+            $onDecoratedInvalid = array_key_exists('decoration_on_invalid', $service) ? $service['decoration_on_invalid'] : 'exception';
+            if ('exception' === $onDecoratedInvalid) {
+                $invalidBehavior = ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE;
+            } elseif ('ignore' === $onDecoratedInvalid) {
+                $invalidBehavior = ContainerInterface::IGNORE_ON_INVALID_REFERENCE;
+            } elseif (null === $onDecoratedInvalid) {
+                $invalidBehavior = ContainerInterface::NULL_ON_INVALID_REFERENCE;
+            } else {
+                throw new InvalidArgumentException(sprintf('The configuration key "decoration_on_invalid" is unsupported for the service "%s" in "%s". Allowed values are "exception", "ignore" and "null".', $id, $file));
```

Comment:
`'Invalid value "%s" for attribute "decoration_on_invalid" on service "%s". Did you mean "exception", "ignore" or null in "%s"?'`

we could also have a dedicated exception message when `'null' === $onDecoratedInvalid`